### ISSUE TITLE
Built-in software USB Hub on PIO SM

### ISSUE THREAD

**Issue by javavi** [No0ne/ps2x2pico#65]
When using the ps2x2pico adapter to simultaneously connect a USB keyboard and mouse, you have to connect them via an external USB hub, which creates additional inconveniences and not all of them are suitable.
Perhaps a good solution would be to use the Pico-PIO-USB library for software implementation of two separate USB-A ports for separate connection of a USB keyboard and USB mouse?
At the same time, the hardware USB port of Pico will remain free for other applications... for example, connecting gamepads with mapping to keyboard buttons.
[https://github.com/sekigon-gonnoc/Pico-PIO-USB](url)

**Comment by No0ne** [comment]
Hi, I'm actually using this already for the ITX-Llama integration: https://github.com/eivindbohler/itxllama/pull/2
Here a patch for TinyUSB is needed to activate both native and PIO USB ports.

I think this would be most suitable with a custom PCB discussed here: https://github.com/No0ne/ps2x2pico/issues/57

**Comment by javavi** [comment]
Isn't it possible to run both PS/2 on one PIO, and USB on another PIO?

**Comment by No0ne** [comment]
Yes thats what I'm doing on the ITX-Llama. There is no PS/2 input/passthru on the ITX-Llama, but the normal ps2x2pico has that feature enabled.

**Comment by No0ne** [comment]
Yes, its already working, one PIO is used for output and one for input. If the input is disabled it can be used for USB as seen on the ITX-Llama.

Do you want to try this out?

**Comment by javavi** [comment]
> Do you want to try this out?

Yes, it would be interesting, where can I see the schematic and firmware?

**Comment by javavi** [comment]
> I don't have any schematic as I don't have any KiCAD skills, but we can agree on using GPIO16/17 and GPIO18/19 as D+/D- and I'll send you a testing firmware build.

OK! I'll try to assemble a prototype of such an adapter device and let you know... and also, will it be possible to use arbitrary pairs of ports for USB interfaces, for example GP0, GP1 and GP2, GP3?

**Comment by No0ne** [comment]
Correct, any GPIO pairs are working (without space between D+/D- and none of the PS/2 pins ;)

**Comment by javavi** [comment]
![2xUSB-PIO-2xPS2](https://github.com/user-attachments/assets/76cbe45d-c26f-4e0e-958c-7fd4f6a4b797)
I plan to do it according to this scheme

**Comment by No0ne** [comment]
Try this: [pio-usb-test.zip](https://github.com/user-attachments/files/17998787/pio-usb-test.zip)

**Comment by javavi** [comment]
Today, I quickly assembled this circuit, flashed your firmware... And yes! It works!
![photo_2024-12-05_01-34-46](https://github.com/user-attachments/assets/df9f8294-f796-4aba-90cd-d77808924be9)

**Comment by javavi** [comment]
Just a question out of curiosity:
Is it possible to do this so that in addition to the ports of the software USB Hub, the built-in USB in Pico also works?

**Comment by javavi** [comment]
I also have an adapter made for standard firmware, but I can solder two USB ports to it.
Could you rebuild your software for this connection scheme?
![2xUSB-PIO-2xPS2_V2](https://github.com/user-attachments/assets/5bb328af-aa92-4a6a-82b0-1b33da67abdc)
* It would be good to have a port configurator for freedom of schematic creativity.

**Comment by No0ne** [comment]
Now all three ports should work: [pio+native-usb-test.zip](https://github.com/user-attachments/files/18022122/pio%2Bnative-usb-test.zip)

**Comment by hax0rbana-adam** [comment]
:wave: I'd be happy to put together a KiCAD schematic and PCB for this. I have a few questions first:

1. Can we slap two PS/2 ports on there in parallel for people who don't have a combo PS/2 port? (just wire all 6 pins to both ports and expect that people will not do something weird like connect them to two computers at once)
2. Do you have a part number for the ZD 3V3 so I don't mess it up?
3. Are those really just 22 ohm resistors?
4. For the PCB, do we want to stick with the Raspberry Pi Pico or Pico 2, or switch to the Waveshare RP2040-zero?

Personally, I like the smaller form factor of the Waveshare and the built-in USB-C port. I'm not super keen on making boards for both the rPi & Waveshare footprints since I'd then feel obligated to build, test and stock them for people who want a ps2x2pico without any soldering.

Oh, and I guess maybe we should move this conversation over to #57 since it's about the hardware and this ticket is about the software?

**Comment by ssokolow** [comment]
> Can we slap two PS/2 ports on there in parallel for people who don't have a combo PS/2 port? (just wire all 6 pins to both ports and expect that people will not do something weird like connect them to two computers at once)

I'd second this. Properly wired PS/2 combo-port breakout cables are surprisingly difficult to find cheaply in Canada so, for the ps2x2pico I built, I just bought a dirt-cheap PS/2 male-to-male cable (Seriously. Why?) from China, cut it in half, and made some little printer-paper-and-scotch/sellotape flags to say which one to plug into the keyboard port and which one to plug into the mouse port.

> For the PCB, do we want to stick with the Raspberry Pi Pico or Pico 2, or switch to the Waveshare RP2040-zero?

As a data point, I like the USB-C port on my YD-RP2040, but a big part of the reason I chose it was that it was the most inexpensive thing I could find. I didn't want to be paying more for the PS/2-to-USB adapter than I paid for the thin client I was repurposing into a fanless retro PC.

**Comment by zoushiyin** [comment]
> Now all three ports should work: [pio+native-usb-test.zip](https://github.com/user-attachments/files/18022122/pio%2Bnative-usb-test.zip)

Which version of the source code does this file correspond to? I would like to make a version. Is that possible?

**Comment by No0ne** [comment]
Its currently in a separate branch: https://github.com/No0ne/ps2x2pico/tree/65-built-in-software-usb-hub-on-pio-sm

**Comment by javavi** [comment]
Look, people are already making adapter boards using a simplified scheme. :))
![photo_2024-12-09_18-38-57](https://github.com/user-attachments/assets/cc2e047a-711e-414c-90f9-b194114304a0)
![photo_2024-12-09_18-40-25](https://github.com/user-attachments/assets/d21fe334-d495-46f1-909a-282bce77f052)
![photo_2024-12-09_18-41-03](https://github.com/user-attachments/assets/02f9bdb6-7d55-4d6f-8d7c-cefa85dea41f)
![photo_2024-12-09_18-41-12](https://github.com/user-attachments/assets/28c16bc4-9087-4bd2-82a0-0d506667c8e2)
https://github.com/xtremespb/usb2ps2-zero

**Comment by zoushiyin** [comment]
> Its currently in a separate branch: https://github.com/No0ne/ps2x2pico/tree/65-built-in-software-usb-hub-on-pio-sm
Hello, @No0ne 

I hope this message finds you well. I have followed the schematic from [this comment](https://github.com/No0ne/ps2x2pico/issues/65#issuecomment-2519971280) to design my PCB for the Pico Zero version. After flashing the firmware, I noticed that there is no response from the inputs and outputs. Additionally, I am unsure where to set the GPIO pins for the USB port, and the LED configuration has not been updated for the Zero version, so the LEDs do not light up.

However, when I flash the latest code, the USB host input for keyboard and mouse works correctly, and the PS/2 output also functions as expected.

**Comment by javavi** [comment]
@No0ne
Everything worked out!
It was just necessary to apply a patch to the library TinyUSB 0.17.0
